pop-api/examples/dao/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+			if description.len() >= STRINGLIMIT.into() {
+				return Err(Error::StringLimitReached);
+			}
+			let proposal = Proposal {
+				description,
+				vote_start: current_block,
+				vote_end,
+				yes_votes: 0,
+				no_votes: 0,
+				executed: false,
+				beneficiary,
+				amount,
+				proposal_id,
+			};
+
+			self.proposals.insert(proposal_id, &proposal);
+
+			self.env()
+				.emit_event(Created { id: proposal_id, creator: caller, admin: contract });
+
+			Ok(())
+		}
+
+		/// Allows Dao's members to vote for a proposal
+		///
+		/// # Parameters
+		/// - `proposal_id` - Identifier of the proposal
+		/// - `approve` - Indicates whether the vote is in favor (true) or against (false) the
+		///   proposal.
+		#[ink(message)]
+		pub fn vote(&mut self, proposal_id: u32, approve: bool) -> Result<(), Error> {
+			let caller = self.env().caller();
+			let current_block = self.env().block_number();
+			let now = self.env().block_number();
+			// let props = self.proposals.clone();
+
+			let mut proposal = self.proposals.get(proposal_id).ok_or(Error::ProposalNotFound)?;
+
+			if current_block < proposal.vote_start || current_block > proposal.vote_end {
+				return Err(Error::VotingPeriodEnded);
+			}
+
+			let member = self.members.get(caller).ok_or(Error::NotAMember)?;
+
+			if member.last_vote >= proposal.vote_start {
+				return Err(Error::AlreadyVoted);
+			}
+
+			if approve {
+				proposal.yes_votes = proposal.yes_votes.saturating_add(member.voting_power);
+			} else {
+				proposal.no_votes = proposal.no_votes.saturating_add(member.voting_power);
+			}
+
+			debug_assert!(proposal.yes_votes > 0);
+
+			self.proposals.remove(proposal_id);
+			self.proposals.insert(proposal_id, &proposal);
+
+			self.members.insert(
+				caller,
+				&Member { voting_power: member.voting_power, last_vote: current_block },
+			);
+
+			self.env().emit_event(Voted { who: Some(caller), when: Some(now) });
+
+			Ok(())
+		}
+
+		/// Enact a proposal approved by the Dao members
+		///
+		/// # Parameters
+		/// - `proposal_id` - Identifier of the proposal
+		#[ink(message)]
+		pub fn execute_proposal(&mut self, proposal_id: u32) -> Result<(), Error> {
+			let vote_end = self.proposals.get(proposal_id).ok_or(Error::ProposalNotFound)?.vote_end;
+
+			// Check the voting period
+			if self.env().block_number() <= vote_end {
+				return Err(Error::VotingPeriodNotEnded);
+			}
+
+			// If we've passed the checks, now we can mutably borrow the proposal
+			// let proposal_id_usize = proposal_id as usize;
+			let mut proposal = self.proposals.get(proposal_id).ok_or(Error::ProposalNotFound)?;
+
+			if proposal.executed == true {
+				return Err(Error::ProposalAlreadyExecuted);
+			}
+
+			if proposal.yes_votes > proposal.no_votes {
+				let contract = self.env().account_id();
+
+				// Execute the proposal
+				let treasury_balance = api::balance_of(self.token_id, contract).unwrap_or_default();
```

Comment:
What happens if there is no token exist? Can simply refactoring to combine with the code at line 250 to throw an inline error.